### ISSUE TITLE
Change Request: Provide a way for rules to know whether they're in fix mode

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint/eslint#17881]
### ESLint version

v8.56.0

### What problem do you want to solve?

Porting https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/issues/115 over to ESLint core: at least one community plugin -`eslint-plugin-expect-type` has a rule whose fixer operates on a separate "snapshot" file in the file system, _not_ the file being linted. That rule has no native way of knowing whether ESLint is being run in `--fix` mode. Because fixers run even if ESLint isn't in fix mode, the fixer can't reliably know whether it should update the file snapshot.

### What do you think is the correct solution?

Two thoughts:

* Can we avoid running `fix()` functions when not in fix mode?
* Alternately, can a rule's `context` object contain info on whether the fixer is being run? Or the `fixer` passed to the `fix()` function?

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

A rules equivalent of what I proposed a year ago in https://github.com/eslint/rfcs/pull/102, perhaps? 🙂

**Comment by mdjermanovic** [comment]
> a rule whose fixer operates on a separate "snapshot" file in the file system, _not_ the file being linted.

I think this is similar to https://github.com/eslint/eslint/issues/16143 (RFC https://github.com/eslint/rfcs/pull/93). It didn't seem feasible because ESLint doesn't always know when the fixes are applied. The issue/RFC was about suggestion fixes, but I think the same applies to auto-fixes.

> * Can we avoid running `fix()` functions when not in fix mode?

`fix()` functions are run even when not in fix mode because the API consumer might want to apply the fixes. For example, in VS Code, when user clicks "Fix this arrow-body-style problem" action, VS Code applies that individual fix, based on the data from [EditInfo](https://eslint.org/docs/latest/integrate/nodejs-api#-editinfo-type) that was provided on the lint message.

> * Alternately, can a rule's `context` object contain info on whether the fixer is being run? Or the `fixer` passed to the `fix()` function?

We could add info about whether ESLint is run in fix mode, but that still doesn't guarantee that the fixes will or will not be applied.

**Comment by JoshuaKGoldberg** [comment]
> that still doesn't guarantee that the fixes will or will not be applied

I think that's fine for rules like `expect-type/expect`. Regardless of whether providing a `fix()` in a rule report means its fixer logic _may_ or _will_ run, we just need a way to make sure that the fixer logic _won't_ run. Knowing to not provide a `fix()` at all would satisfy this use case.

**Comment by mdjermanovic** [comment]
In which cases the fixer logic should run? If I understand https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/issues/115 correctly, the goal is to enable fixes in editors (for example,  "Fix this expect-type/expect problem" popup in VS Code)?

**Comment by JoshuaKGoldberg** [comment]
Hmm, at first I was thinking that the extension could re-run ESLint, or since per https://github.com/eslint/rfcs/pull/93 that seems unlikely, switching to a suggestion instead of a fix. But maybe the right solution is just to always provide a suggestion? And a rule wanting to control when a "fix" is applied generally means that logic should be a suggestion?

**Comment by mdjermanovic** [comment]
Rules don't control if and when either fixes or suggestions will be applied. What use cases currently don't work for `expect-type/expect`?

**Comment by JoshuaKGoldberg** [comment]
Ok, got it - yes, I can verify that both fixes and suggestions are applied immediately. I put a more standalone repro in https://github.com/JoshuaKGoldberg/repros/tree/repro-eslint-expect-type-and-fix. Both trying the rule as-is and switching what's currently a `fix()` to a `suggestion[0].fix()` result in real time updating of the snapshot file in VS Code:

![Screen recording showing modifying an index.ts, and the corresponding index.ts.snap.json updating in real time](https://raw.githubusercontent.com/JoshuaKGoldberg/repros/repro-eslint-expect-type-and-fix/demo.gif)

Now I'm starting to think: maybe the root request here is the ability to mark fixes and  suggestion fixers as having _side effects_? As in, marking that some logic should be run when the fix/suggestion is _executed_, not when _generated_?

Per https://github.com/eslint/rfcs/pull/93#issuecomment-1266454318:

> Actually letting ESLint know that a fix / suggestion was used will be hard. VS Code doesn't provide any callback when a code action is applied. The VS Code ESLint extension basically converts fixes into VS Code's text edits and hands them over to VS Code. The user then selects one from a list and VS Code applies it. All that the extension currently gets is a change event on the document as a result.

I can see a couple potential ways around this:

* Allowing rules to indicate changes to files other than the one being linted
* Sending a feature request to VS Code to provide that _"when a code action is applied"_ callback

**Comment by nzakas** [comment]
The design of rules intentionally doesn't let rules know when they are running in fix mode -- rules are intended to *not* be modal, as they should always do the same thing. Otherwise, we could get hard-to-track-down errors because they occur in fix mode and not in non-fix mode. (There's a lot of isolation designed into rules to prevent whole classes of problems.) So I'm :-1: to letting rules know that ESLint is being run in fix mode.

It seems like the larger issue is much more about fixes being applied to a snapshot rather than the file being linted which, to me, is not a problem that ESLint can or should attempt to solve. Rules were never intended to have filesystem access and definitely not to edit other files in relation to what ESLint is doing. 

I think a better question to ask here is whether what the rule is currently doing actually makes sense? What is the actual problem that it's trying to solve and why is writing a separate file the answer vs. something else?

To that end, it would help to have an explanation of what the rule currently does and why.

**Comment by JoshuaKGoldberg** [comment]
> What is the actual problem that it's trying to solve and why is writing a separate file the answer vs. something else?

The feature in question is [`eslint-plugin-expect-type`](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type)'s [`expect-type/expect` rule](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/blob/ea67df6f3c90b94ac550b7c47f2579b4c18cc46f/docs/rules/expect.md)'s [`$ExpectTypeSnapshot` feature](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/blob/ea67df6f3c90b94ac550b7c47f2579b4c18cc46f/docs/rules/expect.md#expecttypesnapshot). It allows you to place comments like `// $ExpectTypeSnapshot SomeName` in code, then have those comments be turned into snapshots of the corresponding TypeScript type a `__type-snapshots__/*.snap.json` file:

```ts
// file.ts
declare const getTextLength: (text: string) => number;

// $ExpectTypeSnapshot FunctionIdentifier
getTextLength;
```

```jsonc
// __type-snapshots__/file.ts.snap.json
{
	"FunctionIdentifier": "(text: string) => number"
}
```

This is preferred in some situations over inline snapshots ([`// ^?`](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/blob/ea67df6f3c90b94ac550b7c47f2579b4c18cc46f/docs/rules/expect.md#twoslash-syntax-), [`// $ExpectType`](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/blob/ea67df6f3c90b94ac550b7c47f2579b4c18cc46f/docs/rules/expect.md#expecttype)) because TypeScript types can get quite large & verbose. Having a separate snapshot file means they don't take up more than a single line in the source file.

The problem comes from the rule's auto-fixer. When it was first written, it was able to take advantage of ESLint not reading the `text` property of fixes until they're written. At some point that stopped being true: [Running eslint without --fix still updates snapshots _(eslint-plugin-expect-type issue #14)_](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/issues/14). I worked around that no longer being usable in [fix: only update type snapshots if process.argv includes --fix _(eslint-plugin-expect-type PR #113)_](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/pull/113). But the `process.argv.includes("--fix")` solution is a hacky bandaid that doesn't work in all conditions: [🐛 Bug: Type snapshots can't infer ESLint fix mode without a process.argv --fix _(eslint-plugin-expect-type issue #115)_](https://github.com/JoshuaKGoldberg/eslint-plugin-expect-type/issues/115).

**Comment by JoshuaKGoldberg** [comment]
Ah, apologies for the double post, but I came across another use case for applying fixes or suggestions to other files.

CSS-in-JS styling systems such as [Panda CSS](https://panda-css.com) and [Vanilla Extract](https://vanilla-extract.style) will typically expose a function like `css` to generate a CSS class name from an object. Some frameworks allow those can be inline (e.g. Panda CSS) while others require them to be in a `*.css.(j|t)s` file (e.g. Vanilla Extract). Some projects may desire to enforce always putting them in `.css.ts` -file rather than inline and re-evaluated each component render- as a de-cluttering/stylistic preference and/or as a performance optimization.

ESLint plugins for these styling systems will want to be able to enforce only placing `css()` calls in the separate file. Writing a fix/suggestion for those rules would therefore necessitate being able to remove from one file (e.g. `MyComponent.tsx`) and write to another file (e.g. `MyComponent.css.ts`). This is not doable in ESLint right now.

<details>
<summary>Before <code>--fix</code></summary>

```tsx
import { css } from "styled-system/css";

export const MyComponent = () => (
  <p className={css({ color: "foreground" })}>Hello, world!</p>
  //            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
);

**Comment by JoshuaKGoldberg** [comment]
Ah, apologies for the double post, but I came across another use case for applying fixes or suggestions to other files.

CSS-in-JS styling systems such as [Panda CSS](https://panda-css.com) and [Vanilla Extract](https://vanilla-extract.style) will typically expose a function like `css` to generate a CSS class name from an object. Some frameworks allow those can be inline (e.g. Panda CSS) while others require them to be in a `*.css.(j|t)s` file (e.g. Vanilla Extract). Some projects may desire to enforce always putting them in `.css.ts` -file rather than inline and re-evaluated each component render- as a de-cluttering/stylistic preference and/or as a performance optimization.

ESLint plugins for these styling systems will want to be able to enforce only placing `css()` calls in the separate file. Writing a fix/suggestion for those rules would therefore necessitate being able to remove from one file (e.g. `MyComponent.tsx`) **and write to another file** (e.g. `MyComponent.css.ts`). This is not doable in ESLint right now.

<details>
<summary>Before <code>--fix</code></summary>

```tsx
import { css } from "styled-system/css";

export const MyComponent = () => (
  <p className={css({ color: "foreground" })}>Hello, world!</p>
  //            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  // Put css() styles into a ./*.css.ts file. (eslint/eslint-plugin-...)
);
```
</details>

<details>
<summary>After <code>--fix</code></summary>

```tsx
// MyComponent.tsx
import * as styles from "./MyComponent.css.ts";

export const MyComponent = () => (
  <p className={styles.myComponent}>Hello, world!</p>
);
```

```ts
// MyComponent.css.ts
import { css } from "styled-system/css";

export const p = css({
  color: "foreground",
});
```
</details>

**Comment by nzakas** [comment]
Thanks, that context is very helpful.

So, I think writing to a different file from a rule is a dangerous use case -- one of the implicit guarantees of ESLint is that rules won't have side effects. Not only will this not work in runtimes other than Node.js (in the future, we'd like ESLint to run on Deno and in the browser), but there's no way a user could infer that enabling a rule could cause filesystem access outside of the file being linted.

For `expect-type`, fundamentally this doesn't seem like a linting operation at all -- it's code generation. Which, to me, is not what rules are for, and further, not what ESLint is for. This seems like a completely separate process that is only incidentally related to ESLint in that you want to know if the file has changed. And so it seems like what you really need is some kind of utility that can notice when a file has changed and regenerate the associated type snapshots. Does that seem like an accurate description?

**Comment by coderaiser** [comment]
> So, I think writing to a different file from a rule is a dangerous use case -- one of the implicit guarantees of ESLint is that rules won't have side effects...
>
> ...but there's no way a user could infer that enabling a rule could cause filesystem access outside of the file being linted.

What if **Filesystem** was represented as **JavaScript** **AST**? And all manipulations on files was stubbed during development and only works when user provides a flag. In this case implementing of both examples provided by @JoshuaKGoldberg would be just manipulation with **JavaScript** file:

Before:

```js
({
    "filepath": "/",
    "type": "directory"
    "files": [{
        "type": "file",
        "content": "declare const getTextLength: (text: string) => number;",
        "filepath": "/file.ts"
    }]
})
```

After:

```js
({
    "filepath": "/",
    "type": "directory"
    "files": [{
        "type": "file",
        "content": "declare const getTextLength: (text: string) => number;",
        "filepath": "/file.ts"
    }, {
       "type": "directory",
       "filepath": "/__type-snapshots__",
       "files": [{
           "type": "file",
           "filepath": "__type-snapshots__/file.ts.snap.json"
       }]
    }]
})
```

To manipulate files a new **API** can be made to dial with **Filesystem** like: create file, put the data to the file, remove directory etc...

Maybe it is a bit out of scope of ESLint, but the main idea is AST based manipulation - field ESLint operates on all the time.

**Comment by nzakas** [comment]
@coderaiser that's an interesting idea, but I think is overcomplicating the problem.

@JoshuaKGoldberg I'll need to think about this some more. As I said, this is definitely not what rules were designed to support, and I still feel tension between the implicit promise that rules don't have side effects and what you're attempting to do here. At a minimum, I'd want there to be some way for users to know that a rule is doing unexpected things to their filesystem. I know we can't actually prevent people from doing filesystem calls inside of rules but perhaps there could be some blessed way to do so. In any case, I don't think piggybacking on top of the `fix()` method, which is only intended to fix the currently linted file, is the correct approach.

**Comment by mdjermanovic** [comment]
> @mdjermanovic is this true for both fixes and suggestions? It seems like we should provide a flag that people need to pass in order to generate one or both, as that's an unnecessary perf hit if that info isn't being used. (I seem to recall at one point that fixes were only generated when running in fix mode, but maybe that changed?)

Yes, ESLint always generates all fixes and all suggestions since we don't know if they are needed or not.

Related issue: https://github.com/eslint/eslint/issues/14637.

**Comment by Rec0iL99** [comment]
@nzakas do you have an update on this? (https://github.com/eslint/eslint/issues/17881#issuecomment-1885370812) What are the next steps here?

**Comment by nzakas** [comment]
I haven't had any time to revisit exploring this yet. I'm still not sure this is something ESLint should be encouraging, so I'd like more feedback from @eslint/eslint-team as well.

**Comment by nzakas** [comment]
Some prior art:
https://github.com/eslint/eslint/issues/16143

RFC:
https://github.com/eslint/rfcs/pull/93

Clinching RFC comment:
https://github.com/eslint/rfcs/pull/93#issuecomment-1266454318

**Comment by fasttime** [comment]
It seems that [RFC 93](https://github.com/eslint/rfcs/pull/93) was closed because of limitations in the VS Code ESLint extension, one of the "interactive environments" mentioned in the motivation section, which would not have been able to benefit from the change.

Now, the problem of updating file snapshots described here looks like a different use case that would still benefit from the solution proposed in the RFC, and it may be worth to reconsider it in this discussion.

I'm not in favor of adding the ability to apply changes to different files directly in ESLint, because that is a specific requirement that would be hard to generalize in a way to make it useful to other implementations. It would also shift the burden of maintenance for this feature and its many details onto ESLint. A more rationale approach would be to notify rules of a fix and let them take appropriate action.

**Comment by sam3k** [comment]
Per 2024-04-04 TSC meeting, we've decided to table the discussion until we have time to review it more thoroughly.

**Comment by nzakas** [comment]
I was thinking about this some more, and I'm wondering if we are focusing too much on the "write to other files" aspect of this proposal. We may benefit from thinking about this more abstractly. Specifically: what would it mean for a rule to be able to store data in between ESLint sessions?

In both of the examples (`eslint-plugin-expect-type` and `cspell/eslint`), the real goal is to have a database that is built up over time and used over multiple ESLint sessions. The way this is working right now is through saving to files, but I'm curious if saving to files is really the only solution to this problem?

As a thought experiment, what if ESLint provided a persistent key-value store to rules? It would be up to ESLint to determine how best to persist that data, which would mean it could write to files or, if in an environment where filesystem access was not present (i.e. a browser), it could store the info in memory or maybe using `localStorage`.

The question would then become: how necessary is it for users to be able to check-in and version control that data?

**Comment by JoshuaKGoldberg** [comment]
> how necessary is it for users to be able to check-in and version control that data?

For both of those two examples, very necessary _in file system uses_.

* `cspell/eslint`: its `cspell.json` file is a general dictionary used by all [cspell](https://cspell.org) integrations. The lint rule must save to it so CSpell's APIs can pick up on it.
* `eslint-plugin-expect-type`: snapshots should be reviewable alongside their file changes. 

In general for both dictionaries and snapshots, I'd expect to be able to include their changes alongside any other file changes in Git. If a pull request touches something that incurs a change to either, it'll likely need to be reviewed alongside the other changes.

I do like the thought experiment & abstraction of the persistence. One common use case is playgrounds such as eslint.org/play and typescript-eslint.io/play. I could see the playgrounds providing functions that do playground-specific actions instead of file system actions.

Tangentially relevant: https://www.typescriptlang.org/dev/typescript-vfs is a good reference for virtual file systems. We've been wanting to use it on typescript-eslint.io/play for a while.

**Comment by nzakas** [comment]
Hmm okay, back to the drawing board.

**Comment by simonihmig** [comment]
Want to drop one more use case here that is blocked on this: think of a variant of `n/no-extraneaous-import` that is fixable. For example, a common problem we have in a large monorepo with many libraries is that many of those express their dependencies wrongly, by having them listed as `devDependencies` instead of `dependencies`. Which happens to work because of using yarn v1 and hoisting, but it's still bad. A fixable lint rule could handle that, by reporting the bad import, but also fix it optionally by moving the dependency that is imported around in the package.json file. Which btw also shows that the change absolutely needs to happen in a very specific file!

One possibly naive suggestion: when a "fixer" represents the concept of a deferred code transformation, wouldn't it make sense to be able to hook custom "actions" into that? Something like `fixer.addAction(async () => await fs.writeFile('/path/to/file', someData))`. This would only run when the fixer is actually invoked, so wouldn't change the concept of what `fix()` does and what a fixer is used for. And it would also allow to have the fixer do the same AST-based transformation on the linted code parts as it does today, but also handle any kind of other external side effects, but without having to introduce any too specific file-based APIs. 

Does that make sense?

**Comment by nzakas** [comment]
@simonihmig not really. See https://github.com/eslint/eslint/issues/17881#issuecomment-1873439379

**Comment by simonihmig** [comment]
To clarify, given `fixer.addAction(cb)` my expectation would be for `cb` to be only invoked when the fix is actually applied, not when the fix is generated, as ESLint can assume it has side-effects (that's the use case it is made for). No need to discuss names at this point, but for clarity we could also call it `fixer.createEffect(cb)` if that's better :)

**Comment by obrejla** [comment]
Hi, maybe I have another example, when async fixer action could be useful. Imagine something like custom plugin for translations, with a rule checking that a function call of a translation function like `t("My translation key");` uses a translation key defined in e.g. a `en-US.po` file...and if not, it could add such key into the related `.po` file, like adding 
```
msgid "My translation key"
msgstr ""
```
...at the end of the `.po` file. From my PoV this is absolutely valid use case for a custom rule. WDYT?

**Comment by JoshuaKGoldberg** [comment]
**TSC Summary:** This issue suggests expanding suggestions to be able to apply changes to other files. It provides several use cases that cannot be handled by existing systems in ESLint or Git.

**TSC Question:** Would ESLint be interested in an RFC proposing how to solve this?

**Comment by sam3k** [comment]
We've agreed in 2026-01-11 TSC meeting not to pursue applying suggestions to other files. We would suggest opening a separate issue to discuss providing a callback or event instead.

**Comment by JoshuaKGoldberg** [comment]
> separate issue to discuss providing a callback or event instead

@sam3k just to clarify, how would that help in the aforementioned scenarios? I don't know what to request in that issue.

Edit: for future reference, quick Discord link to where the TSC discussed this (I read it and don't follow the applicability): https://discord.com/channels/688543509199716507/688545247843713092/1464008555496603670
